index.js

```diff
@@ -273,61 +281,96 @@ async function handleSensorRequest(currentLocationId, radarType) {
     await redis.addStateMachineData(currentState, currentLocationId)
 
     if (VOIDSTATES.includes(currentState)) {
-      const latestSession = await db.getMostRecentSession(currentLocationId)
-
-      if (typeof latestSession !== 'undefined') {
-        // Checks if no session exists for this location yet.
-        if (latestSession.end_time == null) {
-          // Checks if session is open.
-          await db.updateSessionState(latestSession.sessionid, currentState)
+      let client
+      try {
+        client = await db.beginTransaction()
+        const latestSession = await db.getMostRecentSession(currentLocationId, client)
+
+        if (typeof latestSession !== 'undefined') {
+          // Checks if no session exists for this location yet.
+          if (latestSession.endTime == null) {
+            // Checks if session is open.
+            await db.updateSessionState(latestSession.sessionid, currentState, client)
+          }
+        }
+        await db.commitTransaction(client)
+      } catch (e) {
+        try {
+          await db.rollbackTransaction(client)
+          helpers.log(`handleSensorRequest: Rolled back transaction because of error: ${e}`)
+        } catch (error) {
+          // Do nothing
+          helpers.log(`handleSensorRequest: Error rolling back transaction: ${e}`)
         }
       }
     } else if (TRIGGERSTATES.includes(currentState)) {
-      const client = await db.beginTransaction()
-      const latestSession = await db.getMostRecentSession(currentLocationId, client)
-
-      if (typeof latestSession !== 'undefined') {
-        // Checks if session exists
-        if (latestSession.end_time == null) {
-          // Checks if session is open for this location
-          const currentSession = await db.updateSessionState(latestSession.sessionid, currentState, client)
-          start_times[currentLocationId] = currentSession.start_time
+      let client
+      try {
+        client = await db.beginTransaction()
+        const latestSession = await db.getMostRecentSession(currentLocationId, client)
+
+        if (typeof latestSession !== 'undefined') {
+          // Checks if session exists
+          if (latestSession.endTime == null) {
+            // Checks if session is open for this location
+            const currentSession = await db.updateSessionState(latestSession.sessionid, currentState, client)
+            start_times[currentLocationId] = currentSession.start_time
+          } else {
+            const currentSession = await db.createSession(location.phonenumber, currentLocationId, currentState, client)
+            start_times[currentLocationId] = currentSession.start_time
+          }
         } else {
           const currentSession = await db.createSession(location.phonenumber, currentLocationId, currentState, client)
           start_times[currentLocationId] = currentSession.start_time
         }
```

Comment:
Here in the updateSessionState and createSession functions we're still using the column name and returning a results.rows[0] without running it through createSessionFromRow

Reply:
Good catch! I'll fix those up.
That's what happens when I search for `FROM sessions` and `FROM locations` and forget to search for `RETURNING *` :smile:

Reply:
other than this one, all the other db functions returning raw results.rows are either getCurrentTime, or they're not assigning or doing anything with the return value, so I think they're fine.

Reply:
ah! yeah, my search was for results.rows[0] in db.js, with a space prefix to differentiate it from when we feed it to the object constructor

Reply:
Fixed.